### PR TITLE
init new amp

### PR DESCRIPTION
**Pull request by Jie-Fang**
In the previous mixed precision module, we have to cast input to fp16 first, and store fp16 and fp32 weights in the startup program, and cast fp32 to fp16 weights after updating. Thanks to the black/white list, we don't need to consider master weight copy and the input can maintain fp32. Because we will rewrite the program and insert appropriate cast ops for fp16 execution which will cause inserting cast op for fp16 gradients converting to fp32 in the backward procedure, and updating will occur on fp32 weight. Another advantage is that we don't need to write other functions for saving and loading pretrained fp32 model. Because our parameters are always kept in fp32.

### CODE REVIEW

**Comment by Jie-Fang** [comment]
I tested the some perf between old amp version and this new amp version.

The number of cast op after applying amp

| Model     | old amp | new amp |
| --------- | ------- | ------- |
| ResNet50  | 541     | 114     |
| VGG16     | 71      | 68      |
| GoogleNet | 292     | 288     |

The speed of FP16 and FP32 (sec/step)

| Model     | batch size | old amp                | new amp                |
| --------- | ---------- | ---------------------- | ---------------------- |
| ResNet50  | 128        | FP16: 0.32  FP32: 0.39 | FP16: 0.21  FP32: 0.39 |
| VGG16     | 64         | FP16: 0.16  FP32: 0.29 | FP16: 0.16  FP32: 0.29 |
| GoogleNet | 128        | FP16: 0.11  FP32: 0.12 | FP16: 0.11  FP32: 0.12 |

Because ResNet50 have many batchnorm layer, thus it can reduce the number of cast op dramatically which enables faster speed in FP16.

BTW, the new amp can decrease more GPU memory. For example, The reason for setting batchsize 128 for ResNet50 is that the old amp will exceed GPU memory if setting batchsize 256. While the new amp can set batchsize 256.

**Review by gongweibao — APPROVED** [review]
LGTM
